StyleCop.Analyzers/StyleCop.Analyzers/MaintainabilityRules/SA1402FileMayOnlyContainASingleType.cs

```diff
@@ -0,0 +1,132 @@
+
+        private static readonly DiagnosticDescriptor Descriptor =
+            new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, AnalyzerCategory.MaintainabilityRules, DiagnosticSeverity.Warning, AnalyzerConstants.EnabledByDefault, Description, HelpLink);
+
+        private static readonly Action<SyntaxTreeAnalysisContext, StyleCopSettings> SyntaxTreeAction = HandleSyntaxTree;
+
+        /// <inheritdoc/>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(Descriptor);
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterSyntaxTreeAction(SyntaxTreeAction);
+        }
+
+        private static void HandleSyntaxTree(SyntaxTreeAnalysisContext context, StyleCopSettings settings)
+        {
+            var syntaxRoot = context.Tree.GetRoot(context.CancellationToken);
+
+            var typeNodes = GetTypeDeclarations(syntaxRoot, settings);
+
+            string suffix;
+            var fileName = FileNameHelpers.GetFileNameAndSuffix(context.Tree.FilePath, out suffix);
+            var preferredTypeNode = typeNodes.FirstOrDefault(n => FileNameHelpers.GetConventionalFileName(n, settings.DocumentationRules.FileNamingConvention) == fileName) ?? typeNodes.FirstOrDefault();
+
+            if (preferredTypeNode == null)
+            {
+                return;
+            }
+
+            var foundTypeName = NamedTypeHelpers.GetNameOrIdentifier(preferredTypeNode);
+            var isPartialType = NamedTypeHelpers.IsPartialDeclaration(preferredTypeNode);
+
+            foreach (var typeNode in typeNodes)
+            {
+                if (typeNode == preferredTypeNode || (isPartialType && foundTypeName == NamedTypeHelpers.GetNameOrIdentifier(typeNode)))
+                {
+                    continue;
+                }
+
+                var location = NamedTypeHelpers.GetNameOrIdentifierLocation(typeNode);
+                if (location != null)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
+                }
+            }
+        }
+
+        private static IEnumerable<MemberDeclarationSyntax> GetTypeDeclarations(SyntaxNode root, StyleCopSettings settings)
+        {
+            var typeDeclarations = root.DescendantNodes(descendIntoChildren: node => ContainsTypeDeclarations(node)).OfType<MemberDeclarationSyntax>().ToList();
+            var relevantTypeDeclarations = typeDeclarations.Where(x => IsRelevantType(x, settings)).ToList();
+            return relevantTypeDeclarations;
+        }
+
+        private static bool ContainsTypeDeclarations(SyntaxNode node)
```

Comment:
💭 Consider renaming this to `ContainsTopLevelTypeDeclarations`, as the current name implies that a class containing subclasses should also return true.